encoding/ssz/query/generalized_index.go

```diff
@@ -0,0 +1,358 @@
+package query
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+const (
+	bytesPerChunk = 32
+	bitsPerChunk  = 256
+	listBaseIndex = 2
+)
+
+type Element struct {
+	length  bool
+	name    string
+	indices *[]uint64
+}
+
+// GetGeneralizedIndexFromPath calculates the generalized index for a given path.
+// To calculate the generalized index, two inputs are needed:
+// 1. The sszInfo of the root info, to be able to navigate the SSZ structure
+// 2. The path to the field (e.g., "field_a.field_b[3].field_c")
+// It walks the path step by step, updating the generalized index at each step.
+func GetGeneralizedIndexFromPath(info *sszInfo, path []PathElement) (uint64, error) {
+	if info == nil {
+		return 0, errors.New("sszInfo is nil")
+	}
+
+	// If path element list is empty, no generalized index can be computed.
+	if len(path) == 0 {
+		return 0, errors.New("cannot compute generalized index for an empty path")
+	}
+
+	// Starting from the root generalized index
+	root := uint64(1)
+	currentInfo := info
+
+	for _, pathElement := range path {
+		name := pathElement.Name
+		element, err := processPathElement(name)
+		if err != nil {
+			return 0, err
+		}
+
+		// If we descend to a basic type, the path cannot continue further
+		if isBasicType(currentInfo.sszType) {
+			return 0, fmt.Errorf("cannot descend into basic type %s for path element %q", currentInfo.sszType, name)
+		}
+
+		// Check that we are in a container to access fields
+		if currentInfo.sszType != Container {
+			return 0, fmt.Errorf("indexing requires a container field step first, got %s", currentInfo.sszType)
+		}
+
+		// Check if path element contains an array index (e.g., field_name[5])
+		var idx *uint64
+		if element.indices != nil && len(*element.indices) > 0 {
+			// Note: shortcut, extend to multi-dimensional arrays later
+			idx = &(*element.indices)[0]
+		}
+
+		// Retrieve the field position and SSZInfo for the field in the current container
+		fieldPos, fieldSsz, err := getContainerFieldByName(currentInfo, element.name)
+		if err != nil {
+			return 0, fmt.Errorf("container field %q not found: %w", element.name, err)
+		}
+
+		// Get the chunk count for the current container
+		chunkCount, err := getChunkCount(currentInfo)
+		if err != nil {
+			return 0, fmt.Errorf("chunk count error: %w", err)
+		}
+
+		// Update the generalized index to point to the specified field
+		root = updateRoot(root, 1, chunkCount, fieldPos)
+		currentInfo = fieldSsz
+
+		// Check if a path element is a length field
+		if element.length {
+			// Length field is only valid for List and Bitlist types
+			if fieldSsz.sszType != List && fieldSsz.sszType != Bitlist {
+				return 0, fmt.Errorf("len() is only supported for List and Bitlist types, got %s", fieldSsz.sszType)
+			}
+			currentInfo = &sszInfo{sszType: UintN, fixedSize: 8}
+			root = updateRoot(root, 1, 2, 1)
```

Comment:
```suggestion
			root = root*2 + 1
```

I think we can simply put the calculation here